### PR TITLE
Mark return type of an unresolved method during load as required

### PR DESCRIPTION
**Pull request by JohnDumbell**
If you have an instance of an interface whose class can't be resolved (built via reflection) and attempt to call a method on it, then you also need to mark the return type of this method as being required. 

Explicitly in this case there was a class being returned from an interface that had an Enum field, and this Enum wasn't having its clinit run and the special $VALUES field being set up because during the load it couldn't work out what instance of class was hiding behind the interface.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: e39ec95ac19a925bb47e92244f6747290e8f9c4b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93190021
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5097c7c2a7d27f3a26a5c52a73d8b98957e0d3b6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93190285

**Review by smowton — APPROVED** [review]

File: jbmc/regression/jbmc/NondetEnumOpaqueReturn/indirect/IndirectNestedEnumOnReturn.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package indirect;
+
+public class IndirectNestedEnumOnReturn {
+    public void main() throws ClassNotFoundException, IllegalAccessException, InstantiationException {
+        PrimaryInterface primary = (PrimaryInterface)Class.forName("PrimaryInstance").newInstance();
+        ClassWithEnumField enumClass = primary.getEnumOwningClass();
```
Comment (antlechner) [inline jbmc/regression/jbmc/NondetEnumOpaqueReturn/indirect/IndirectNestedEnumOnReturn.java]: :no_entry_sign: I can't reproduce the invariant violation on develop with this test. There should be an additional line that accesses the enum, e.g.
```
        enumClass.enumInstance.isActive();
```
(where `enumInstance` would have to be public).

**Review by antlechner — DISMISSED** [review]
The code changes look good but the test is incomplete.
Will we have to make a similar change for parameter types of methods with no known callees?

**Comment by JohnDumbell** [comment]
@antlechner We could for safety, but the assumption that if you're passing something _in_ you will have an instance of it somewhere is relatively reliable, unless we start calling stubs within stubs. I left it as-is because other code seems to make this assumption.

File: jbmc/regression/jbmc/NondetEnumOpaqueReturn/indirect/IndirectNestedEnumOnReturn.java (hunk 1 above)
Comment (JohnDumbell) [reply jbmc/regression/jbmc/NondetEnumOpaqueReturn/indirect/IndirectNestedEnumOnReturn.java]: It's not really to do with the enum being called - it's specifically that a class instance with an enum field is a return value and not recognised as required anywhere else and then being nondet initialised. But it certainly isn't throwing the violation as it should, so I'll fix it up.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0608020f894ff98e156846221839d4021f4e0f28).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94798195

**Comment by JohnDumbell** [comment]
This issue is only directly seen in another project because we are a bit more eager with stubbing than default JBMC. The issue exists, but can't be directly replicated here, so it was suggested we add a test elsewhere that asserts this, and otherwise put this change in so other people can benefit (if they come up against it)

@antlechner Where the tests I sent you OK?

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 2]
```diff
@@ -253,17 +253,41 @@ bool ci_lazy_methodst::handle_virtual_methods_with_no_callees(
     if(!candidate_target_methods.empty())
       continue;
 
+    const java_method_typet &java_method_type =
+      to_java_method_type(virtual_function_call.type());
```
Comment (allredj) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp:257]: :pick: I prefer having definitions closer to where there are used (i.e. line 268)

**Review by allredj — APPROVED** [review]
Looks fine to me.

**Comment by allredj** [comment]
@antlechner Are you aware of an occurrence of this bug on TG?

**Comment by antlechner** [comment]
@allredj Not aware of this having affected TG but we do still get a few invariant violations on our benchmarks so I will check if any of them went away after the next TG pointer bump.

**Pull request merged**